File: src/main/java/io/vproxy/vfx/util/algebradata/ColorData.java
package io.vproxy.vfx.util.algebradata;

import io.vproxy.vfx.util.FXUtils;
import javafx.scene.paint.Color;

public class ColorData implements AlgebraData<ColorData> {
    private Color color;
    private final float h;
    private final float s;
    private final float b;
    private final double alpha;

    public ColorData(Color color) {
        this.color = color;
        var hsb = FXUtils.toHSB(color);
        h = hsb[0];
        s = hsb[1];
        b = hsb[2];
        alpha = color.getOpacity();
    }

    public ColorData(float h, float s, float b, double alpha) {
        this.h = h;
        this.s = s;
        this.b = b;
        this.alpha = alpha;
    }

    public ColorData(float[] hsb, double alpha) {
        h = hsb[0];
        s = hsb[1];
        b = hsb[2];
        this.alpha = alpha;
    }

    public Color getColor() {
        if (color == null) {
            double alpha = this.alpha;
            if (alpha > 1) {
                alpha = 1;
            } else if (alpha < 0) {
                alpha = 0;
            }
            color = FXUtils.fromHSB(h, s, b, alpha);
        }
        return color;
    }

    @Override
    public ColorData plus(ColorData other) {
        return new ColorData(h + other.h, s + other.s, b + other.b, alpha + other.alpha);
    }

    @Override
    public ColorData minus(ColorData other) {
        return new ColorData(h - other.h, s - other.s, b - other.b, alpha - other.alpha);
    }

    @Override
    public ColorData multiply(double v) {
        return new ColorData((float) (h * v), (float) (s * v), (float) (b * v), alpha * v);
    }

    @Override
    public ColorData dividedBy(double v) {
        return new ColorData((float) (h / v), (float) (s / v), (float) (b / v), alpha / v);
    }
}

File: src-ts/src/util/algebradata/ColorData.ts
import { AlgebraData } from './AlgebraData.js';
import { Color, toHSB, fromHSB } from '../../javafx/color.js';

export class ColorData implements AlgebraData<ColorData> {
  private _color: Color | null;
  private readonly h: number;
  private readonly s: number;
  private readonly b: number;
  private readonly alpha: number;

  constructor(color: Color);
  constructor(h: number, s: number, b: number, alpha: number);
  constructor(hsb: [number, number, number], alpha: number);
  constructor(arg0: Color | number | [number, number, number], sOrAlpha?: number, b?: number, alpha?: number) {
    if (arg0 instanceof Color) {
      this._color = arg0;
      const hsb = toHSB(arg0);
      this.h = hsb[0];
      this.s = hsb[1];
      this.b = hsb[2];
      this.alpha = arg0.opacity;
    } else if (Array.isArray(arg0)) {
      this._color = null;
      this.h = arg0[0];
      this.s = arg0[1];
      this.b = arg0[2];
      this.alpha = sOrAlpha ?? 1;
    } else {
      this._color = null;
      this.h = arg0;
      this.s = sOrAlpha ?? 0;
      this.b = b ?? 0;
      this.alpha = alpha ?? 1;
    }
  }

  get color(): Color {
    if (this._color === null) {
      let a = this.alpha;
      if (a > 1) a = 1;
      else if (a < 0) a = 0;
      this._color = fromHSB(this.h, this.s, this.b, a);
    }
    return this._color;
  }

  plus(o: ColorData): ColorData {
    return new ColorData(this.h + o.h, this.s + o.s, this.b + o.b, this.alpha + o.alpha);
  }
  minus(o: ColorData): ColorData {
    return new ColorData(this.h - o.h, this.s - o.s, this.b - o.b, this.alpha - o.alpha);
  }
  multiply(v: number): ColorData {
    return new ColorData(this.h * v, this.s * v, this.b * v, this.alpha * v);
  }
  dividedBy(v: number): ColorData {
    return new ColorData(this.h / v, this.s / v, this.b / v, this.alpha / v);
  }
}
